core/shared/src/main/scala/zio/stm/STM.scala

```diff
@@ -558,6 +558,10 @@ object STM {
                 Sync(globalLock) {
                   if (isValid(journal)) commitJournal(journal) else loop = true
                 }
+              } else {
+                Sync(globalLock) {
+                  if (isInvalid(journal)) loop = true
```

Comment:
I'm surprised this fixes anything, because above, we already have that `analysis ne JournalAnalysis.Invalid`. In other words, it may fix something only non-deterministically, because it happens to be checking "again" to see if it became invalid (and more will be invalid if checked again, so you'll occasionally fix a race issue this way).

Could be the real fix is a bit deeper. 

What are your thoughts?

Reply:
Scenario:
One more try to convince you:
- please take a look at test, only `r0` and `r1` exists. STM that reads them both and other STM that updates them both, racing.
- `stm` being processed here is read-only only - it won't contain `Entry` that `isChanged`.
- the other, updating, `STM` on same `TRef`s, it's commiting at this moment - `r0` has been committed and thread pauses before committing `r1`.
- hence `value = stm.exec(journal)` assigns `Succeed(1)` to `value` and adds `r0` and `r1` to `journal`.
- `val analysis = analyzeJournal(journal)` assigns `ReadOnly` to `analysis` <- I think it's possible, because `stm.exec` started after other STM commited `r0`.
- now the other STM continues to commit and updates `r1.versioned` with `1`
- our `stm` enters `if` in L522, then without my fix it breaks the loop, having `value == Succeed(1)`
- if my fix was in place validity of `journal` would be checked again in critical section - not running concurrently with other thread commiting `TRef`s

Reply:
My proposal adds check of journal validity for ReadOnly journal **inside critical section**.
Journal check outside critical section can only decide that journal is invalid, but for decision that journal is valid, it has to be done inside critical section.